src/Features/LanguageServer/Protocol/Handler/InlayHint/InlayHintHandler.cs

```diff
@@ -0,0 +1,112 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.InlineHints;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.InlayHint
+{
+    [ExportCSharpVisualBasicStatelessLspService(typeof(InlayHintHandler)), Shared]
+    [Method(Methods.TextDocumentInlayHintName)]
+    internal sealed class InlayHintHandler : ILspServiceDocumentRequestHandler<InlayHintParams, LSP.InlayHint[]?>
+    {
+        private readonly IGlobalOptionService _optionsService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public InlayHintHandler(IGlobalOptionService optionsService)
+        {
+            _optionsService = optionsService;
+        }
+
+        public bool MutatesSolutionState => false;
+
+        public bool RequiresLSPSolution => true;
+
+        public TextDocumentIdentifier GetTextDocumentIdentifier(InlayHintParams request)
+            => request.TextDocument;
+
+        public async Task<LSP.InlayHint[]?> HandleRequestAsync(InlayHintParams request, RequestContext context, CancellationToken cancellationToken)
+        {
+            var document = context.GetRequiredDocument();
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textSpan = ProtocolConversions.RangeToTextSpan(request.Range, text);
+
+            var inlineHintService = document.GetRequiredLanguageService<IInlineHintsService>();
+            var options = _optionsService.GetInlineHintsOptions(document.Project.Language);
+
+            var hints = await inlineHintService.GetInlineHintsAsync(document, textSpan, options, displayAllOverride: false, cancellationToken).ConfigureAwait(false);
+            if (hints.IsEmpty)
+            {
+                return Array.Empty<LSP.InlayHint>();
+            }
+
+            using var _ = ArrayBuilder<LSP.InlayHint>.GetInstance(out var inlayHints);
+            var syntaxVersion = await document.GetSyntaxVersionAsync(cancellationToken).ConfigureAwait(false);
+            var inlayHintCache = context.GetRequiredLspService<InlayHintCache>();
+
+            // Store the members in the resolve cache so that when we get a resolve request for a particular
+            // member we can re-use the inline hint.
+            var resultId = inlayHintCache.UpdateCache(new InlayHintCache.InlayHintCacheEntry(hints, request.TextDocument, syntaxVersion));
+
+            for (var i = 0; i < hints.Length; i++)
+            {
+                var hint = hints[i];
+                var linePosition = text.Lines.GetLinePosition(hint.Span.Start);
+                var kind = hint.Ranking == 0.0
```

Comment:
what is this?

Reply:
the ranking? it was used to fix an editor issue with two intratextadornmenttags occupying the same buffer. we assigned 0.0 to parameters and 1.0 to type hints.